babeld: add ubus bindings

PolynomialDivision opened the pull request:
A better integration of babeld with OpenWrt is to connect the daemon to the ipc. So far, we can only communicate via a websocket. With ubus we can send and receive commands in json format.

adschm commented on babeld/src/ubus.h:
```diff
@@ -0,0 +1,10 @@
+
+int babeld_ubus_isset_ubus(fd_set* readfds);
+
+void babeld_ubus_receive();
```
missing newline at EOF ...

adschm commented on babeld/src/ubus.h:
```diff
@@ -0,0 +1,10 @@
+
```
this empty line can be dropped?

PolynomialDivision commented:
@adschm Thanks a lot for all the comments already. :) I added now the first part of the "dump" to show "my-id". But acutally, my id is always changing?

```
root@OpenWrt:/tmp# ubus call babeld dump
{
	"my-id": 4334176
}
root@OpenWrt:/tmp# ubus call babeld dump
{
	"my-id": 4334204
}
root@OpenWrt:/tmp# ubus call babeld dump
{
	"my-id": 4334232
}
root@OpenWrt:/tmp# ubus call babeld dump
{
	"my-id": 4334260
}
root@OpenWrt:/tmp# ubus call babeld dump
{
	"my-id": 4334176
}
root@OpenWrt:/tmp# ubus call babeld dump
{
	"my-id": 4334204
}
```
Any idea? I will have a deeper look into this.

PolynomialDivision commented:
Lol, the id is a string. I will fix everxthing and mention you again, then I'm finished.

PolynomialDivision commented:
The only thing that I'm still not sure is that I would like to check folr `v4mapped(prefix)` and divide into v4 and v6 addresses. But not sure how to do that.

PolynomialDivision commented:
I'm adding that list stuff.

kerneis commented:
This seems like a significant addition rather than a mere patch. I wonder if it would make sense to try and upstream it, or at least ask @jech how he feels about it. Also, ubus.h and ubus.cc should really be part of the patch file rather than added directly to the sources in my opinion.

PolynomialDivision commented:
@kerneis It is a very common procedure to add ubus bindings to OpenWrt in that way.
A famous example ist hostapd:
https://github.com/openwrt/openwrt/tree/master/package/network/services/hostapd



> Also, ubus.h and ubus.cc should really be part of the patch file rather than added directly to the sources in my opinion.

I would point to some drivers, or again the hostapd. I think having such a big patchfile is not that easy. And maintainance is better in seperated files.

>  I wonder if it would make sense to try and upstream it,

I would keep it in OpenWrt, since no other OS has ubus bindings. But if @jech want, I can add this also upstream. :)

I hope that noone is against adding ubus bindings. I added only basic functionality and want to add more stuff. :) I think this will enable great features and new opportunities. (Soon, I will publish more code)

PolynomialDivision commented:
@adschm I think I'm finished. I'm sorting now ipv4 and ipv6.
I have now the calls:

```
root@OpenWrt:~# ubus call babeld get_info
{
	"babeld-version": "babeld-1.9.2",
	"my-id": "32:xx:xx:xx:xx:xx:a3:7d",
	"host": "OpenWrt"
}
```

```
root@OpenWrt:~# ubus call babeld get_routes
{
	"IPv4": {
		"10.2.0.1/32": {
			"src-prefix": "0.0.0.0/0",
			"route_metric": 96,
			"route_smoothed_metric": 96,
			"refmetric": 0,
			"id": "62:e3:xx:xx:xx:fb:4e:a6",
			"seqno": 41381,
			"channels": "",
			"age": 17,
			"via": "fe80::xx:xxff:fxx:4ea6",
			"nexthop": " nexthop ",
			"installed": true,
			"feasible": true
		}
	"IPv6": {
		
	}
}

```
root@OpenWrt:~# ubus call babeld get_xroutes
{
	"IPv4": {
		"10.0.0.3/32": {
			"src-prefix": "0.0.0.0/0",
			"metric": 0
		},
		"10.0.0.0/24": {
			"src-prefix": "0.0.0.0/0",
			"metric": 0
		}
	},
	"IPv6": {
		"fdfa:xx:xx::1/128": {
			"src-prefix": "::/0",
			"metric": 0
		}
	}
}

```

PolynomialDivision commented:
Argh, I have to rename all that neighbor stuff... ^^

PolynomialDivision commented:
@kerneis Are u okay, if I create a separate package, e.g. ubabeld? Then I can maintain the stuff there and go further with it. And at the end we can find a way to merge again?

jech commented:
Thanks for the work.

> This seems like a significant addition rather than a mere patch. I wonder if it
> would make sense to try and upstream it, or at least ask @jech how he feels
> about it.

I'd rather not accept it upstream, since then I would have to maintain it,
and I no longer have any devices able to run recent releases of OpenWRT
(all my OpenWRT devices have 32MB RAM).

I agree with Gabriel that this is a fairly significant addition.  However,
I find it clean and modular, and wouldn't object to it getting into
OpenWRT.  Please add a description of your changes somewhere, so that
people aren't confused about the divergence.

I have no objection to the fork being called babeld.  I have no objection
to the fork being called ubabeld.  On the other hand, I would like to
avoid the situation in which there are both babeld and ubabeld in OpenWRT,
as that would lead to confusion.

I'm happy to trust Gabriel with making the right decision.

PolynomialDivision commented:
> Thanks for the work.

> However, I find it clean and modular, and wouldn't object to it getting into OpenWRT.

Thanks it is always nice to hear that people are thankful for your contribution! :D Thanks for making babel. :)

I will rewrite my commit message to add a better description.

PolynomialDivision commented:
@kerneis anything new? ;)

kerneis commented:
Based on what @jech said I think it's fine to merge keeping the name babeld. Ideally, the version information (`babeld -V`) would be updated to mention the patch, so that it's easier to know what fork is running when asking users for debug information.

I'd like to review the patch more carefully before it's merged, but won't have time before next week. In the meantime, if somebody can test (I don't have access to OpenWrt devices these days, which probably means I shouldn't maintain this package anymore), I'd be grateful for feedback. @zorun maybe?

Gabriel

PolynomialDivision commented:
> the version information (`babeld -V`) would be updated to mention the patch, so that it's easier to know what fork is running when asking users for debug information.

Okay. I think I will just add something like
https://github.com/jech/babeld/blob/master/generate-version.sh#L13
`echo "#define BABELD_VERSION \"$version\"-ubus-mod"`

PolynomialDivision commented:
```
root@OpenWrt:/tmp# ubus call babeld get_info
{
	"babeld-version": "babeld-1.9.2-ubus-mod",
	"my-id": "32:b5:c2:ff:fe:6e:a3:7d",
	"host": "OpenWrt"
}
```

zoobab commented:
What does ""channel": -2," means in the example?

PolynomialDivision commented:
> What does ""channel": -2," means in the example?

I think that is a babel magic number for weird conection, but I have no idea. Just guessing.

jech commented:
Yes.  https://github.com/jech/babeld/blob/master/interface.h#L85

PolynomialDivision commented:
Something new?
I really would like to go on. :P

kerneis commented on babeld/src/ubus.c:
```diff
@@ -0,0 +1,357 @@
+    .n_methods = ARRAY_SIZE(babeld_methods),
+};
+
+static void ubus_add_oject() {
```
typo: ubus_add_object

kerneis commented on babeld/src/ubus.c:
```diff
@@ -0,0 +1,357 @@
+}
+
+int babeld_ubus_add_read_sock(fd_set *readfds, int maxfd) {
+  FD_SET(ctx->sock.fd, readfds);
```
Check that ctx is not null.

kerneis commented on babeld/src/ubus.c:
```diff
@@ -0,0 +1,357 @@
+static void ubus_add_oject() {
+  int ret;
+
+  ret = ubus_add_object(ctx, &babeld_object);
```
Can you please include the relevant header, I couldn't find it?

PolynomialDivision replied:
that is #include <libubus.h>

PolynomialDivision replied:
so it is already included in ubus.h, so I won't double include it in ubus.c?

jech commented on babeld/src/ubus.c:
```diff
@@ -0,0 +1,357 @@
+
+#include "ubus.h"
+
+LIST_HEAD(xroute_ipv4_list);
```
It was difficult to write, there's no reason it should be easy to read.

PolynomialDivision commented on babeld/patches/600-add-ubus.patch:
```diff
@@ -0,0 +1,63 @@
+     change_smoothing_half_life(4);
+     has_ipv6_subtrees = kernel_has_ipv6_subtrees();
+ 
++    babeld_add_ubus();
```
You mean an option on babeld start?

kerneis replied:
Yes, something similar to the `-g` flag in the regular babeld.

PolynomialDivision replied:
Any idea why this does not work?
https://github.com/openwrt-routing/packages/pull/630/files#diff-d414ab8ec5eed4328cb50c6a16058041e6842cc0821536ced6c61fc3a34673ffR93-R94

kerneis replied:
I think you test the value before parsing it in if(ubus_bindings)  babeld_add_ubus().

PolynomialDivision replied:
shit. ^^ thanks!!! :)

PolynomialDivision commented on babeld/src/ubus.c:
```diff
@@ -0,0 +1,357 @@
+  void *prefix;
+  char channels[100];
+
+  const unsigned char *nexthop =
```
that's babel code and not documented :/ not sure why it is needed.

PolynomialDivision commented on babeld/src/ubus.c:
```diff
@@ -0,0 +1,357 @@
+  struct neighbour *neighbour;
+};
+
+static struct ubus_context *ctx;
```
making the `blob_buf b` global is a typical thing in openwrt. But I can make it local.

PolynomialDivision commented:
@kerneis Thanks for all the good comments! I changed everything.

PolynomialDivision commented:
rebased on current master

PolynomialDivision commented:
I added
```
	# option 'ubus_bindings' 'false'
```
to the config.

kerneis commented:
I will try to give this another round of review quickly (github interface doesn't make it easy, sorry). In the meantime, could anybody else confirm that the latest patch compiles and behaves correctly please? I don't have openwrt devices at home anymore…

kerneis commented on babeld/patches/600-add-ubus.patch:
```diff
@@ -0,0 +1,97 @@
+ extern int max_request_hopcount;
+ 
++// config option for ubus bindings
++extern int ubus_bindings;
```
Please move this to ubus.h instead: it's easier to think of exported variables in terms of translation units, and to have the declaration of a variable in the header corresponding to the .cc file where it is defined. (Also, this will remove babel.h completely from this patch, which will make it easier to read).

kerneis commented on babeld/src/ubus.c:
```diff
@@ -0,0 +1,373 @@
+
+#include "ubus.h"
+
+// config option for ubus bindings
```
// Whether to enable ubus bindings (boolean option).

kerneis commented on babeld/src/ubus.c:
```diff
@@ -0,0 +1,373 @@
+// config option for ubus bindings
+int ubus_bindings = 0;
+
+// struct definition for sorting xroutes
```
Are we really sorting them? It looked to me we just separate v4 and v6.

// List of exported routes (to be used with ubox's list helpers)

idem below

kerneis commented on babeld/src/ubus.c:
```diff
@@ -0,0 +1,373 @@
+  struct neighbour *neighbour;
+};
+
+// pointer to ubus context
```
When commenting, try to focus on the why rather than paraphrasing the type information (eg. here, why is the variable global and what is its lifetime):

// Shared state maintained throughout calls to handle ubus messages.

kerneis commented on babeld/src/ubus.c:
```diff
@@ -0,0 +1,373 @@
+  if (!gethostname(host, sizeof(host)))
+    blobmsg_add_string(&b, "host", host);
+
+  ret = ubus_send_reply(ctx, req, b.head);
```
I'm confused: do you intend to use ctx or ctx_local here? Maybe they're just the same? Because your initial code used the local one, and since the rename it uses the local one.

Maybe it makes more sense to call the local variables ctx and the global one shared_ctx or persistent_ctx?

PolynomialDivision replied:
Yes. This was an mistake. But actually it makes no difference as you already explained. :)

kerneis commented on babeld/src/ubus.c:
```diff
@@ -0,0 +1,373 @@
+// pointer to ubus context
+static struct ubus_context *ctx;
+
+static void babeld_ubus_babeld_info(struct ubus_context *ctx_local,
```
// Sends a babel info message on ubus socket.

kerneis commented on babeld/src/ubus.c:
```diff
@@ -0,0 +1,373 @@
+    fprintf(stderr, "Failed to send reply: %s\n", ubus_strerror(ret));
+}
+
+static void babeld_add_xroute_buf(struct xroute *xroute, struct blob_buf *b) {
```
// Appends an exported route message entry to the buffer.


idem for similar functions below

kerneis commented on babeld/src/ubus.c:
```diff
@@ -0,0 +1,373 @@
+  blobmsg_close_table(b, prefix);
+}
+
+static void babeld_ubus_get_xroutes(struct ubus_context *ctx_local,
```
// Sends an exported route message on ubus socket, splitting apart IPv4 and IPv6 routes.

idem for similar functions below

PolynomialDivision replied:
"exported routes", or?

kerneis commented on babeld/src/ubus.c:
```diff
@@ -0,0 +1,373 @@
+    .n_methods = ARRAY_SIZE(babeld_methods),
+};
+
+static bool ubus_init_object() {
```
// Registers handlers for babel methods in the global ubus context.

kerneis commented on babeld/src/ubus.c:
```diff
@@ -0,0 +1,373 @@
+  return true;
+}
+
+static bool babeld_ubus_init(void) {
```
// Initializes the global ubus context, connecting to the bus to be able to receive and send messages.

kerneis reviewed (changes requested):
A couple of minor changes still needed, but overall it seems ready for inclusion to me. Thanks for the great work. If nobody else chimes in, I'll merge your PR once the changes are done, and revert if someone complains that it breaks them.

PolynomialDivision commented:
> A couple of minor changes still needed, but overall it seems ready for inclusion to me. Thanks for the great work. If nobody else chimes in, I'll merge your PR once the changes are done, and revert if someone complains that it breaks them.

Thanks for all your efforts. :)

If this is finished, I would like to send the received routes also over the bus. Then I can monitor easily and directly if someone else 
tries to announce certain prefixes. I think I have to do this here https://github.com/jech/babeld/blob/master/route.c#L859
If this patch is merged I can poll for new updates, but with the above mentioned changes I can react to changes more quickly. :)

jech commented:
What is the performance impact of all of that?

kerneis commented:
> What is the performance impact of all of that?

I'm not worried too much currently because it's disabled by default, and it has to be polled so even if active the user decides what the cost is. It doesn't look more expensive than the babel local interface (at least asymptotically, it copies quite a lot and I don't know what the overhead is of the ubox function calls).

kerneis commented:
> If this is finished, I would like to send the received routes also over the bus. Then I can monitor easily and directly if someone else
> tries to announce certain prefixes. I think I have to add some function to the "local_notify..." parts.
> If this patch is merged I can poll for new updates, but with the above mentioned changes I can react to changes more quickly. :)

Two points:
- when working on github pull requests, please create follow-up commits rather than amending, and then squash at the very end of the process. We shouldn't need to do this (gerrit for instance handles the amend pattern very well), but github garbage collects old versions of the patch, so it's impossible to compare changes since the last review if you amend, and it's painful as a reviewer.
- I'd rather have you open a separate bug or pull request to discuss the pushing model. I'm not sure how you intend to do it, because you'll need to push the initial state when a new client starts listening on the bus. Is this something ubox allows?

PolynomialDivision commented:
> > What is the performance impact of all of that?
> 
> I'm not worried too much currently because it's disabled by default, and it has to be polled so even if active the user decides what the cost is. It doesn't look more expensive than the babel local interface (at least asymptotically, it copies quite a lot and I don't know what the overhead is of the ubox function calls).

This is also my opinion.

For me, it makes things more efficient because I can request only the things I need. The parsing can then also be done very efficiently in c. The only thing I find unnecessary in the code is that when I separate in ipv4 and ipv6 to makes two lists. Sadly, my testbed is not that big right now, but with increasing number of nodes I will look again at performance.

Overall, I think to connect babeld to ubus will make it more attractive to other openwrt-devs, since they are used to use this way of inter process communication. :)

This will not be the last commit you will receive from me. ;) Probably, we will role out the babel mesh daemon over [Berlin Freifunk Network](https://berlin.freifunk.net/), since it is working very smooth in our testbed. :)

> when working on github pull requests, please create follow-up commits rather than amending, and then squash at the very end of the process. We shouldn't need to do this (gerrit for instance handles the amend pattern very well), but github garbage collects old versions of the patch, so it's impossible to compare changes since the last review if you amend, and it's painful as a reviewer.

Okay. :)

> push the initial state when a new client starts listening on the bus.

I have to look into the code to see what u mean with that. I will write some code over the weekend and make a new PR. ;)
